### PR TITLE
Ensure panic safety for all FFI methods

### LINKED ISSUE #32: Use panic-safety helper functions from ffi-support
Panics in the rust code are not currently handled very well, and I suspect would actually trigger undefined behaviour by propagating across the FFI boundary.  We should use the support functions offered by [ffi-support](https://docs.rs/ffi-support/0.4.0/ffi_support/index.html) to ensure safe panic handling in all cases.

The most basic thing that could work:

* All ffi functions should have an `ExternError` out parameter
* The rust scaffolding for all standalone functions and object methods should wrap the call in [`call_with_output`](https://docs.rs/ffi-support/0.4.0/ffi_support/fn.call_with_output.html).

Object methods could instead rely on the `call` helpers of the `HandleMap` for the same effect, but I think it would be worth our while to catch and handle panics in argument conversion or other operations that might occur before we call into the `HandleMap` at all.

┆Issue is synchronized with this [Jira Task](https://jira.mozilla.com/browse/SYNC-1556)

### PR DESCRIPTION
**Pull request by linacambridge**
This commit changes all FFI functions to take an `out_err` param, so
that they're called using `call_with_result`. Any functions that don't
have a `[Throws]` annotation in the IDL will throw a generic
`InternalError` instead.

Closes #32.

### CODE REVIEW

**Comment by linacambridge** [comment]
I think Python is broken now because methods without `[Throws]` are now expecting a `RustError` argument as well, but that'll be fixed with #215.

**Comment by rfk** [comment]
Will there be any conflicts or problems if https://github.com/mozilla/uniffi-rs/pull/215 lands ahead of this PR, or should we hold on of that one until this one is landed?

File: uniffi_bindgen/src/bindings/swift/templates/ErrorTemplate.swift
Comment (rfk) [inline uniffi_bindgen/src/bindings/swift/templates/ErrorTemplate.swift:22]: Possibly contrary opinion: I'm not convinced it's valuable to define an elaborate series of cases for each individual type of internal error that might be thrown. If anyone ever finds themselves matching on these different cases and doing anything other than freaking out and filing a bug against uniffi, something has gone terribly wrong 😅.

In that spirit, the most important thing is that these errors surface with a useful human-readable message that helps us find and fix the bug. If enumerating all the cases helps us achieve that, awesome. But if they're just adding layers of indirection to a fixed string value, then I worry they might set incorrect expectations about how these internal errors should be handled.

I'm not going to r- for that though, just my 2c.

**Comment by tarikeshaq** [comment]
@rfk I think we can land #215. And after this merges, we can spin off another issue to tackle panics in Python based on what we come with here! Thoughts?

> Naively, as a consumer I think I would expect a panicking [Throws(E)] function to throw an InternalError.Panic, rather than throwing e.g. some sort of E.InternalError.Panic variant. Does that mesh with what's being proposed here?

Hmm ya this makes more sense... Although I can't instantly think of a clean way to do this (since we pass the error type as a part of the `rustCall`).

This makes me wonder if we should just `panic` on the side of the binding (`fatalError` in swift and maybe a `RuntimeException` in Kotlin) if we receive an internal error (as opposed to raising a special exception)

I defer to @linacambridge's judgement though 😅

File: uniffi_bindgen/src/bindings/swift/templates/ErrorTemplate.swift
Comment (linacambridge) [reply uniffi_bindgen/src/bindings/swift/templates/ErrorTemplate.swift:22]: Hmm, thinking some more about this, I'm coming around to what you and Tarik suggested—assert for any kind of internal error instead of bubbling it up as an exception. In real-world code, `[Throws]` is for defining error cases that are meaningfully recoverable—a network error or duplicate experiment name, for example. A Rust-side panic _might_ be, but...how do you fix that? If the app were fully in Rust, it would crash. The only reason we catch panics is to avoid stack unwinding into the foreign language, not as a graceful error recovery mechanism.

**Comment by rfk** [comment]
>  And after this merges, we can spin off another issue to tackle panics in Python based on what we come with here! Thoughts?

Yep. I'll probably try to tackle that as part of my big make-python-work refactor over in #214 (and I'd hate to derail any critical-path nimbus stuff in the service of the python bindings).

**Comment by linacambridge** [comment]
Yes, let's land #215 first, and then this. I'll go ahead and make the changes to fatally assert on an internal error, instead of throwing.

File: uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt
Comment (linacambridge) [inline uniffi_bindgen/src/bindings/kotlin/templates/ErrorTemplate.kt:96]: OK, I simplified the error hierarchy here so that we have a catch-all "internal error" instead of getting fancy with reporting panic and invalid handle codes. We might want to report these specially to Sentry, for example, but I totally agree consumers shouldn't be matching on them. WDYT?

**Comment by linacambridge** [comment]
I went ahead and did a cleanup pass and added a quick test—would you mind taking another look please, @tarikeshaq? ☺️

File: examples/arithmetic/tests/bindings/test_arithmetic.kts
Comment (rfk) [inline examples/arithmetic/tests/bindings/test_arithmetic.kts:22]: IIUC this will catch the `RuntimeException` thrown above, meaning the test will always pass even if the rust code didn't panic at all. Should we catch a more specific type of error here?

File: examples/arithmetic/tests/bindings/test_arithmetic.py
Comment (rfk) [inline examples/arithmetic/tests/bindings/test_arithmetic.py:28]: This will catch and ignore the `AssertionError` from "should have panicked" above. In this case you can use the `try-else` syntax like this:

```
try:
    div(8, 0)
except:
    pass
else:
    assert False, "Should have panicked"

File: uniffi_bindgen/src/bindings/python/templates/ErrorTemplate.py
Comment (rfk) [inline uniffi_bindgen/src/bindings/python/templates/ErrorTemplate.py:19]: nit: please throw a `RuntimeException` here rather than the generic `Exception` base class.

**Review by rfk — APPROVED** [review]
Looks good to me :-)

**Pull request merged**
